stringify.go

```diff
@@ -0,0 +1,220 @@
+						s += fmt.Sprintf("%s=%s\n", variable, value)
+					}
+					s += "\n"
+				}
+			} else if fieldStruct.Type.Elem().Kind() == reflect.Ptr && fieldStruct.Type.Elem().Elem().Kind() == reflect.Struct {
+				iter := fieldValue.MapRange()
+				for iter.Next() {
+					res, err := stringifyStructFields(iter.Value().Elem())
+					if err != nil {
+						return "", err
+					}
+
+					if iter.Key().String() == "" {
+						s += fmt.Sprintf("[%s]\n", iniFieldName)
+					} else {
+						s += fmt.Sprintf("[%s \"%s\"]\n", iniFieldName, iter.Key())
+					}
+					s += res + "\n"
+				}
+			} else {
+				return "", fmt.Errorf("The map values must either be of string or *struct type, instead they are of %s type", fieldStruct.Type.Elem().Kind())
+			}
+		}
+	}
+
+	return s, nil
+}
+
+func stringifyStructFields(value reflect.Value) (string, error) {
+	if value.Kind() != reflect.Struct {
+		return "", fmt.Errorf("Expected a struct type from %v, but instead got %s\n", value, value.Kind())
+	}
+
+	var s string
+	for i := 0; i < value.NumField(); i++ {
+		fieldValue := value.Field(i)
+		fieldStruct := value.Type().Field(i)
+		if !fieldStruct.IsExported() {
+			continue
+		}
+
+		if fieldStruct.Tag.Get("gcfg") == "extra_values" {
+			if fieldStruct.Type != reflect.TypeOf(map[string]string{}) && fieldStruct.Type != reflect.TypeOf(map[string][]string{}) {
+				return "", fmt.Errorf("Expected either a map[string]string or map[string][]string type, but instead got %s\n", fieldStruct.Type)
+			}
+
+			iter := fieldValue.MapRange()
+			for iter.Next() {
+				key := iter.Key()
+				value := iter.Value()
+				if value.Kind() == reflect.String {
+					s += fmt.Sprintf("%s=%s\n", key, value)
+				} else {
+					for j := 0; j < value.Len(); j++ {
+						s += fmt.Sprintf("%s=%s\n", key, value.Index(j))
+					}
+				}
+			}
+		} else {
+			if err := iterateMaybeSlice(fieldValue, func(innerValue reflect.Value) error {
+				res, err := iniValue(innerValue)
+				if err != nil {
+					return err
+				}
+				s += fmt.Sprintf("%s=%s\n", iniKey(fieldStruct), res)
+				return nil
+			}); err != nil {
+				return "", err
+			}
+		}
+	}
+
+	return s, nil
+}
+
+// This function implements the inversion of `fieldFold`.
+func iniKey(fieldStruct reflect.StructField) string {
+	tag := newTag(fieldStruct.Tag.Get("gcfg"))
+	if tag.ident != "" {
+		return tag.ident
+	}
+
+	name := fieldStruct.Name
+	firstRune, firstRuneSize := utf8.DecodeRuneInString(name)
+	if firstRune == 'X' {
+		secondRune, _ := utf8.DecodeRuneInString(name[firstRuneSize:])
+		if unicode.IsLetter(secondRune) && !unicode.IsLower(secondRune) && !unicode.IsUpper(secondRune) {
+			name = name[firstRuneSize:]
+		}
+	}
+
+	return strings.ToLower(strings.Replace(name, "_", "-", -1))
+}
+
+// This function implements the inversion of `setters`.
+func iniValue(value reflect.Value) (string, error) {
+	if bigIntValue, ok := value.Interface().(big.Int); ok {
+		return bigIntValue.String(), nil
+	}
```

Comment:
Can we just check if the value implements `.String()` and use that? 

```golang
if stringer, ok := value.Interface().(fmt.Stringer); ok {
     return stringer.String(), nil
}
```

Reply:
Ya, that would work too, but it would break the order used by https://github.com/please-build/gcfg/blob/v1/set.go#L64-L66 when setting the value. I'll add a comment to clarify.